src/libraries/Microsoft.PowerFx.Interpreter/Functions/LibraryMutation.cs

```diff
@@ -0,0 +1,279 @@
+            {
+                return arg0;
+            }
+
+            if (args[1] is not RecordValue baseRecord)
+            {
+                return args[1];
+            }
+
+            var dvalue = MutationProcessUtils.MergeRecords(args.Skip(2));
+
+            if (!dvalue.IsValue)
+            {
+                return dvalue.ToFormulaValue();
+            }
+
+            // If base record is {}, then collect.
+            if (baseRecord.IsEmptyRecord)
+            {
+                return (await tableValue.AppendAsync(dvalue.Value, cancellationToken).ConfigureAwait(false)).ToFormulaValue();
+            }
+            else
+            {
+                return (await tableValue.PatchAsync(baseRecord, dvalue.Value, cancellationToken).ConfigureAwait(false)).ToFormulaValue();
+            }
+        }
+    }
+
+    // If arg1 is pure PFx record, it will return a runtime internal error.
+    // Patch(DS, record_with_keys_and_updates)
+    internal class PatchSingleRecordImpl : PatchSingleRecordFunction, IAsyncTexlFunction5
+    {
+        public async Task<FormulaValue> InvokeAsync(IServiceProvider runtimeServiceProvider, FormulaType irContext, FormulaValue[] args, CancellationToken cancellationToken)
+        {
+            MutationProcessUtils.ThrowIfPFxV1NotActive(runtimeServiceProvider.GetService<Features>(), "Patch");
+
+            var arg0 = args[0];
+
+            if (args[0] is LambdaFormulaValue arg0lazy)
+            {
+                arg0 = await arg0lazy.EvalAsync().ConfigureAwait(false);
+            }
+
+            if (arg0 is not TableValue tableValue)
+            {
+                return arg0;
+            }
+
+            if (args[1] is not RecordValue recordValue)
+            {
+                return args[1];
+            }
+
+            return (await tableValue.PatchAsync(recordValue, cancellationToken).ConfigureAwait(false)).ToFormulaValue();
+        }
+    }
+
+    // Patch(DS, table_of_rows, table_of_updates)
+    internal class PatchAggregateImpl : PatchAggregateFunction, IAsyncTexlFunction5
+    {
+        public async Task<FormulaValue> InvokeAsync(IServiceProvider runtimeServiceProvider, FormulaType irContext, FormulaValue[] args, CancellationToken cancellationToken)
+        {
+            MutationProcessUtils.ThrowIfPFxV1NotActive(runtimeServiceProvider.GetService<Features>(), "Patch");
+
+            var arg0 = args[0];
+
+            if (args[0] is LambdaFormulaValue arg0lazy)
+            {
+                arg0 = await arg0lazy.EvalAsync().ConfigureAwait(false);
+            }
+
+            if (arg0 is not TableValue tableValue)
+            {
+                return arg0;
+            }
+
+            var arg1Rows = ((TableValue)args[1]).Rows;
+            var arg2Rows = ((TableValue)args[2]).Rows;
+
+            // !!!TODO Do we need to check if args[1] and args[2] counters are equal?
+            // PA does not give any error if they are not equal.
```

Comment:
Currently Power Apps doesn't validate it - it effectively takes the smaller of the lengths and goes with it. Returning an error seems appropriate in this case.

Reply:
I'll address this internally.

Reply:
PFxV1 will try to match the args sizes.